### ISSUE TITLE
Ability to time individual hooks

### ISSUE THREAD

**Issue by johnbillion** [johnbillion/query-monitor#309]
Via @crstauf on https://github.com/johnbillion/query-monitor/pull/301#issuecomment-364177323:

Possible to add `qm/start`/`qm/stop` onto a hook, and if there's no param (function name) provided, assume `current_action()`?

QM could enable this via a wrapper function such as `qm_start()` and `qm_stop()` which would then be used like this:

```php
add_action( 'after_setup_theme', 'qm_start', 0 );
add_action( 'after_setup_theme', 'qm_stop', 999 );
```

Maybe even only require the start action and QM adds the stop action itself.

**Comment by johnbillion** [comment]
Yeah QM makes use of negative integers itself, usually `-9999` for early hooks. It doesn't guarantee being the first, but it covers most cases.

The internal timer could maybe check to see if another action is hooked on an earlier priority and warn you. I'll have a play around.

**Comment by crstauf** [comment]
I've started on a PR for timing hooks. Here's the code I tested with:

```php
add_action( 'temp', 'qm_start',    0 );
add_action( 'temp', 'qm_lap',      9 );
add_action( 'temp', 'qm_lap',     11 );
add_action( 'temp', 'qm_stop',  9999 );

add_action( 'temp', function() { sleep( 2 ); } );
add_action( 'init', function() { do_action( 'temp' ); } );
```

The result is this:

<img width="1078" alt="screen shot 2019-02-15 at 10 56 12 pm" src="https://user-images.githubusercontent.com/4573033/52894266-edf90c80-3174-11e9-8c71-79e480e4ed2f.png">

I can't seem to figure out though, `QM_Backtrace`, and how to get the component right. I've tried placing the code in different places (another plugin, mu-plugin, theme) and passing different values to `QM_Backtrace::ignore()` by calling `QM_Collector_Timing->start[ current_action() ]->get_trace()->ignore( $x )`, but after I get past the `Plugin: query-monitor` component, its identifies the component as core after that, I think due to the use of `do_action`.

@johnbillion Can you please advise on how to have QM select and display the correct component?

**Comment by crstauf** [comment]
Related: #260

**Comment by crstauf** [comment]
Related: #340
